Providing more detail on EndpointSlice implementation (#18343)
kubernetes/website: content/en/docs/concepts/services-networking/endpoint-slices.md

{{< feature-state for_k8s_version="v1.17" state="beta" >}}

_EndpointSlices_ provide a simple way to track network endpoints within a
Kubernetes cluster. They offer a more scalable and extensible alternative to

> Endpoints.

Reviewer: In a way, it's kind of handy that there isn't already a concept page for Endpoint.

{{% capture body %}}

## Motivation

The Endpoints API has provided a simple and straightforward way of
tracking network endpoints in Kubernetes. Unfortunately as Kubernetes clusters
and Services have gotten larger, limitations of that API became more visible.
Most notably, those included challenges with scaling to larger numbers of
network endpoints.

Since all network endpoints for a Service were stored in a single Endpoints
resource, those resources could get quite large. That affected the performance
of Kubernetes components (notably the master control plane) and resulted in
significant amounts of network traffic and processing when Endpoints changed.
Endpoint Slices help you mitigate those issues as well as provide an extensible
platform for additional features such as topological routing.

## Endpoint Slice resources {#endpointslice-resource}

> In Kubernetes, an Endpoint Slice contains references to a set of network

Reviewer 2: I prefer `EndpointSlice`, see [documentation style guide's standards](https://kubernetes.io/docs/contribute/style/style-guide/#documentation-formatting-standards).

Author: Thanks for pointing that out! I'd missed that in the style guide, fixed now.

      topology.kubernetes.io/zone: us-west2-a
```

By default, EndpointSlices managed by the EndpointSlice controller will have no
more than 100 endpoints each. Below this scale, EndpointSlices should map 1:1
with Endpoints and Services and have similar performance.

EndpointSlices can act as the source of truth for kube-proxy when it comes to
how to route internal traffic. When enabled, they should provide a performance
improvement for services with large numbers of endpoints.

### Address Types

EndpointSlices support three address types:

* IPv4
* IPv6
* FQDN (Fully Qualified Domain Name)

### Topology

> Within EndpointSlices, each endpoint can contain relevant topology information.

Reviewer: ```suggestion
Each endpoint within an EndpointSlices can contain relevant topology information.
```
?

Author: That does read much better, thanks!

* `topology.kubernetes.io/region` - The region this endpoint is in.

The values of these labels are derived from resources associated with each
endpoint in a slice. The hostname label represents the value of the NodeName
field on the corresponding Pod. The zone and region labels represent the value
of the labels with the same names on the corresponding Node. 

### Management

By default, EndpointSlices are created and managed by the EndpointSlice
controller. There are a variety of other use cases for EndpointSlices, such as
service mesh implementations, that could result in other entities or controllers
managing additional sets of EndpointSlices. To ensure that multiple entities can
manage EndpointSlices without interfering with each other, a
`endpointslice.kubernetes.io/managed-by` label is used to indicate the entity
managing an EndpointSlice. The EndpointSlice controller sets
`endpointslice-controller.k8s.io` as the value for this label on all
EndpointSlices it manages. Other entities managing EndpointSlices should also
set a unique value for this label.

### Ownership

In most use cases, EndpointSlices will be owned by the Service that it tracks

> endpoints for. This is indicated by an owner reference on each EndpointSlice as

Reviewer: Aside: a glossary entry for _owner reference_ would be nice to have.

Author: Completely agree, I was looking for one as I writing this.

well as a `kubernetes.io/service-name` label that enables simple lookups of all
EndpointSlices belonging to a Service.

## EndpointSlice Controller

The EndpointSlice controller watches Services and Pods to ensure corresponding
EndpointSlices are up to date. The controller will manage EndpointSlices for
every Service with a selector specified. These will represent the IPs of Pods
matching the Service selector.

### Size of EndpointSlices

By default, EndpointSlices are limited to a size of 100 endpoints each. You can
configure this with the `--max-endpoints-per-slice` {{< glossary_tooltip
text="kube-controller-manager" term_id="kube-controller-manager" >}} flag up to
a maximum of 1000.

### Distribution of EndpointSlices

Each EndpointSlice has a set of ports that applies to all endpoints within the
resource. When named ports are used for a Service, Pods may end up with
different target port numbers for the same named port, requiring different
EndpointSlices. This is similar to the logic behind how subsets are grouped
with Endpoints.

> The Controller tries to fill EndpointSlices as full as possible, but does not

Reviewer: nit:
```suggestion
The controller tries to fill EndpointSlices as full as possible, but does not
```
because Controller is not an API object. You could also write “control plane” here.

(and later changes likewise)

Author: Good catch, thanks! I think I caught all the other times as well.

actively rebalance them. The logic of the Controller is fairly straightforward:

1. Iterate through existing EndpointSlices, remove endpoints that are no longer
   desired and update matching endpoints that have changed.
2. Iterate through EndpointSlices that have been modified in the first step and
   fill them up with any new endpoints needed.
3. If there's still new endpoints left to add, try to fit them into a previously
   unchanged slice and/or create new ones.
   

> Importantly, the third step prioritizes limiting EndpointSlice updates over a

Reviewer: I wonder: is this too much detail for a concept page?

(I can add this detail into PR #16885 instead)

Author: Yeah, this is potentially getting into too much detail here. I really like the concept of your PR, especially if we can link from resource specific docs to their controller logic prominently.

Author: Open to adding it all in this PR and then you pulling it out in your PR, or leaving it for you to add in your PR. Whatever is best for you.

Reviewer: > Whatever is best for you
Same. It's fine to have an idea about future tweaks and those don't block merging this PR (IMO).

Author: Yep, agreed. I'd say getting more information in the docs sooner is probably a win, so leaving it in here makes sense now, with the idea that it will eventually move to your PR.

      topology.kubernetes.io/zone: us-west2-a
```

By default, Endpoint Slices managed by the Endpoint Slice controller will have no
more than 100 endpoints each. Below this scale, Endpoint Slices should map 1:1
with Endpoints and Services and have similar performance.

Endpoint Slices can act as the source of truth for kube-proxy when it comes to
how to route internal traffic. When enabled, they should provide a performance
improvement for services with large numbers of endpoints.

### Address Types

Endpoint Slices support three address types:

* IPv4
* IPv6
* FQDN (Fully Qualified Domain Name)

### Topology

With Endpoint Slices, each endpoint can contain relevant topology information.
Endpoint Slices managed by the Endpoint Slice controller will contain the

> following information when it's available:

Reviewer: Are these labels per-slice or per-endpoint?

(My inferred learning outcome: the reader should be clear that if 2 endpoints have _different_ topology information, that implies 2 EndpointSlice resources in the API server).

Reviewer: I would reference https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/ to clarify that this topology information is conveyed using _labels_, and that these names are meaningful outside of the context of Services and their endpoints.

Author: I've updated this to clarify the topology information is stored per endpoint, as well as providing more information about the source of these values. These values aren't all derived from labels, and they also aren't necessarily labels, although the validation is similar. Do the updates make sense? Is there a good place to reference the labels, annotations, and taints doc you suggested above?

* `kubernetes.io/hostname` - The name of the Node this endpoint is on.
* `topology.kubernetes.io/zone` - The zone this endpoint is in.
* `topology.kubernetes.io/region` - The region this endpoint is in.

> ### Management

Reviewer: Management and ownership?

(Do any controllers set `ownerRef`? That might be implied; if so, I would spell it out).

Author: Good catch, I ended up adding a new ownership section below as that ended up being an important omission. Thanks!

By default, Endpoint Slices will be created and managed by the Endpoint Slice
controller, but they can be useful beyond that. Endpoint Slices may be useful
for a variety of additional use cases such as service mesh implementations. To
ensure that multiple entities can manage Endpoint Slices without interfering
with each other, a `endpointslice.kubernetes.io/managed-by` label is used to
indicate the entity managing an Endpoint Slice. The Endpoint Slice controller
sets `endpointslice-controller.k8s.io` as the value for this label on all
Endpoint Slices it manages. Other entities managing Endpoint Slices should also
set a unique value for this label.

> ## Endpoint Slice Controller

Reviewer: I'd write this as
```sugggestion
## EndpointSlice controller
```

Author: fixed, thanks!

The Endpoint Slice controller watches Services and Pods to ensure corresponding
Endpoint Slices are up to date. The controller will manage Endpoint Slices for 

> every Service with a Selector specified. These will represent the IPs of Pods

Reviewer: Is “Selector” an API object of the kind that you can `kubectl some-verb Selector`?

If not, I'd write it lower case. That way, readers can get a clearer understanding of what nouns Kubernetes have.
It is OK to use a glossary tooltip for the first use of a term with a very specific Kubernetes meaning.

Author: Updated, thanks! I've also added a bunch of glossary tooltips throughout, let me know if it's too much...

matching the Service Selector. 

### Size of Endpoint Slices

> By default, Endpoint Slices will be limited to a size of 100 endpoints each. This

Reviewer: If you write this paragraph in the active voice, it kind-of forces you to tell the reader what imposes the limit (which can be as broad as “Kubernetes”). I recommend active voice here.

Author: Updated, thanks!

can be configured with the `--max-endpoints-per-slice` Controller Manager flag
up to a maximum of 1000.

### Distribution of Endpoint Slices

Each Endpoint Slice has a set of ports that applies to all endpoints within the
resource. When named ports are used for a Service, Pods may end up with
different target port numbers for the same named port, requiring different
Endpoint Slices. This is similar to the logic behind how subsets are grouped
with Endpoints.

> The Controller tries to fill Endpoint Slices as full as possible, but does not

Reviewer: I have a pull request (#16885) in draft to add a set of reference pages for all built-in control loops. If / when that lands, this level of detail could migrate there. What do you think?

actively rebalance them. The logic of the Controller is fairly straightforward:

1. Iterate through existing Endpoint Slices, remove endpoints that are no longer
   desired and update matching endpoints that have changed.
2. Iterate through Endpoint Slices that have been modified in the first step and
   fill them up with any new endpoints needed.
3. If there's still new endpoints left to add, try to fit them into a previously
   unchanged slice and/or create new ones.
   
Importantly, the third step prioritizes limiting Endpoint Slice updates over a
perfectly full distribution of Endpoint Slices. As an example, if there are 10
new endpoints to add and 2 Endpoint Slices with room for 5 more endpoints each,
this approach will create a new Endpoint Slice instead of filling up the 2
existing Endpoint Slices. In other words, a single Endpoint Slice creation is

> preferrable to multiple Endpoint Slice updates.

Reviewer 2: A suggestion about this part, maybe we could take some typical examples for explaining the scales logical,  a good example is [pod qos](https://github.com/kubernetes/website/commit/f83de3f601bf853e92bf742236c3a331d226cadb#diff-43e0a44ed6b2437d8a5d7480fc2b58f0R100).

Author: I like that idea, good examples are always helpful when I'm trying to learn something new. I've been struggling with how to communicate that well with EndpointSlices where the scale is so significant. YAML snippets generally don't seem to work here because they'd get really long pretty quickly here, do you have any ideas for how I could write examples better here?

Reviewer 2: Yes, no need to write examples by yaml, I think something like [this](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#example-states), use list of items to explain the cases and the behavior of EndpointSlice should be clear enough for our users.

> With Kube Proxy running on each Node and watching Endpoint Slices, every change

Reviewer: “kube-proxy”, so:
```suggestion
With kube-proxy running on each Node and watching EndpointSlices, every change
```

Author: Fixed, thanks!
